include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -659,14 +662,47 @@ class cuda_vm_resource_base : public memory_resource<memory_kind::device> {
       to_map.push_back({ next_unmapped, block_idx });
     }
 
-    // Prevent `get_free_blocks` from unmapping blocks that are already mapped to this range.
-    mark_as_unavailable(ptr, size);
+    // a list of blocks to unmap
+    SmallVector<std::pair<va_region*, int>, 256> to_unmap;
+    {
+      // Prevent `get_free_blocks` from unmapping blocks that are already mapped to this range.
+      available_backup_ = region->available;
+      int num_avail_backup = region->available_blocks;
+      mark_as_unavailable(ptr, size);
+
+      // Let's find some free blocks which we can reuse.
+      get_free_blocks(to_unmap, blocks_to_map);
 
+      // Restore original availability mask
+      region->available = available_backup_;
+      region->available_blocks = num_avail_backup;
+    }
+
+    int new_blocks = blocks_to_map - to_unmap.size();
     SmallVector<cuvm::CUMem, 256> free_blocks;
-    get_free_blocks(free_blocks, blocks_to_map);
-    assert(static_cast<int>(free_blocks.size()) == blocks_to_map);
+    free_blocks.resize(blocks_to_map);
+    for (int i = to_unmap.size(); i < blocks_to_map; i++) {
+      // This can throw, but we should leave the object in a consistent state
```

Comment:
```suggestion
      // This can throw, but should leave the object in a consistent state
```

Reply:
I meant:
This can throw, but (_this_) should leave the object in a consistent state ;)

Reply:
I know that is why `we` sounds strange here.

Reply:
It doesn't matter - both are fine and it's just an internal comment - not even doxygen.